man/getxrec.3
.TH GETXREC 3 "12 September 1995"
.FB
.SH NAME
getxrec \- read an indexed record
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_getxrec(key, db)
.br
char *key;
.br
fb_database *db;
.br
.PP
.SH DESCRIPTION
As explained in the \fIdatabase\fP(4) manual page,
a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
\fIFb_getxrec\fP attempts to load
a database record which has a current index entry of \fIkey\fP into \fIdb\fP.
A binary search is done on the
current index in an attempt to locate a record with
an index entry of \fIkey\fP. In the case of flat \fIFirstBase\fP indexes,
if no record is found, the index
overflow area is searched sequentially.
.PP
For alpha type fields, \fIkey\fP needs to be left justified in a field of
blanks, and for numeric keys, \fIkey\fP should be right
justified. Use \fImakess\fP(3) to generate a search string key.
For example:
.PP
.nj
.nf
.ft CW
      fb_field *f;

      f = cdb_kp[0];
      fb_fetch(f, cdb_afld, cdb_db);
      fb_makess(cdb_afld, f->type, f->size);
      fb_getxrec(cdb_afld, cdb_db);
.ft
.fi
.ju
.PP
Once a record number is retrieved from the index, \fIfb_getxrec\fP uses
\fIgetrec\fP(3) to load the record into the database structures.
.PP
When using a \fIFirstBase\fP flat index, 
the physical position number of the index entry, the index record number,
is stored in the database structure \fIdb->bsrec\fP.
If \fIdb->bsrec\fP is greater
than \fIdb->bsend\fP, the returned record's key is stored in the overflow area
of the index, otherwise it is stored in the binary searchable area.
.PP
However, when using a \fIFirstBase\fP Btree+ index, the \fIdb->bsrec\fP value
is actually the record number (not the index record number).
.PP
When \fIfb_getxrec\fP fails to locate an entry, the record that is loaded will
be the one where the search failure occurred. Again, on an FB_ERROR return from
\fIfb_getxrec\fP, the record loaded will be where the search key would have
been, the record containing a key value greater than the search key.
If such a record does not exist, then \fIdb->rec\fP
and \fIdb->bsrec\fP will be set to 0, indicating an undefined record.
.PP
Note: Before calling \fIfb_getxrec\fP on a normal (flat) \fIFirstBase\fP index,
the index header numbers should
be updated using \fIgetxhead\fP if there have been new records added
to the index since it was opened, as in the following code fragment:
.PP
.nj
.nf
.ft CW
      /* get bsmax, bsend from index header file */
      fb_getxhead(dp->ihfd, &dp->bsmax, &dp->bsend);
.ft
.fi
.ju
.PP
.SH CLIENT/SERVER
The \fIfb_getxrec\fP routine will transparently
call the \fIfb_getxrec_clnt\fP mechanism
when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP is set to one.
See \fIfb_clnt_create\fP(3) for more details.
.PP
.SH SEE ALSO
getxhead(3), nextxrec(3), makess(3),
getrec(3), getirec(3), opendb(3), database(4), btree(4), index(4).
.SH DIAGNOSTICS
\fIFb_getxrec\fP will return FB_AOK on a successful call, and FB_ERROR when
the record could not be found or loaded.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/getxrec.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: getxrec.c,v 9.0 2001/01/09 02:56:59 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Getxrec_sid[] = "@(#) $Id: getxrec.c,v 9.0 2001/01/09 02:56:59 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

extern long fail_irec;
extern short int search_position;
extern short int cdb_use_rpc;
extern short int cdb_secure;
extern long cdb_failrec;
extern short int cdb_loadfail;

#if FB_PROTOTYPES
static adjustidx(fb_database *db);
#else
static adjustidx();
#endif /* FB_PROTOTYPES */

extern long fb_megasearch(int fd, char *key, int col, long bsbeg, long bsend,
   long bsmax, int recsiz, int backup, char *buf);
extern long fb_btree_search(char *key, fb_bidx *bi, fb_bseq *bs);
extern long fb_key_eval(fb_bseq *bs);

/*
 *  getxrec - force a record with a particular index entry
 *	to be loaded into dp or return FB_ERROR.
 *	dp->bsrec is set to index record number for standard index
 *	dp->bsrec is set to *actual* fb_database record number for btree+ index
 */

   fb_getxrec(s, dp)
      char *s;
      fb_database *dp;
      
      {
         long rec;
	 int st = FB_ERROR;

#if RPC
         if (cdb_use_rpc)
            return(fb_getxrec_clnt(s, dp));
#endif /* RPC */
         dp->rec = dp->bsrec = 0L;
         cdb_failrec = 0L;
         /* b_tree handled if needed */
         if (dp->b_tree){
            rec = fb_btree_search(s, dp->b_idx, dp->b_seq);
            if (rec > 0 && rec <= dp->reccnt){
               dp->bsrec = rec;
               if (fb_getrec(rec, dp) != FB_ERROR)
                  st = FB_AOK;
               if (cdb_secure && fb_record_permission(dp, READ) == FB_ERROR)
                  st = FB_ERROR;
               }
            else
               dp->bsrec = cdb_failrec;
            }
         /* standard FB index */
         else{
	    rec = fb_megasearch(dp->ifd, s, 0, 1, dp->bsend, dp->bsmax,
               dp->irecsiz, 1, dp->irec);
            if (rec > 0 && rec <= dp->bsmax){
               dp->bsrec = rec;
               rec = atol((char *) 
                  (dp->irec + dp->irecsiz - 11));	/* FB_RECORDPTR + 1 */
               if (rec > 0 && rec <= dp->reccnt)
                  if (fb_getrec(rec, dp) != FB_ERROR)
                     st = FB_AOK;
                  if (cdb_secure && fb_record_permission(dp, READ) == FB_ERROR)
                     st = FB_ERROR;
               }
            else
               dp->bsrec = fail_irec;
	    }
         if (st == FB_ERROR && cdb_loadfail)
            adjustidx(dp);
	 return(st);
      }

/*
 * adjustidx - loads the failure point of the record
 */

   static adjustidx(db)
      fb_database *db;

      {
         int st = FB_ERROR;
         long max;

         if (db->b_tree)
            max = db->reccnt;
         else
            max = db->bsmax;

         if (db->bsrec > 0L && db->bsrec <= max){
            if (!db->b_tree){			/* standad index */
               if (search_position == 1)
                  st = fb_nextxrec(db);
               else
                  st = fb_currentxrec(db);
               }
            else				/* btree index */
               st = fb_getrec(db->bsrec, db);
            }
         if (st == FB_ERROR)
            db->rec = db->bsrec = 0;
      }
